test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -1142,58 +1138,103 @@ arbitraryActionWithVars _ ctx _st = QC.frequency $ concat [
         Open{} -> ()
         Duplicate{} -> ()
 
-    findBlobRefsVars :: [Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))]
-    findBlobRefsVars = fmap fromLookupResults vars1 ++ fmap fromQueryResults vars2
+    genTableVar = QC.elements tableVars
+
+    tableVars :: [Var h (WrapTableHandle h IO k v blob)]
+    tableVars =
+      [ fromRight v
+      | v <- findVars ctx Proxy
+      , case lookupVar ctx v of
+          MEither (Left _)                  -> False
+          MEither (Right (MTableHandle th)) ->
+            Map.member (Model.tableHandleID th) (Model.tableHandles st)
+      ]
+
+    genCursorVar = QC.elements cursorVars
+
+    cursorVars :: [Var h (WrapCursor h IO k v blob)]
+    cursorVars =
+      [ fromRight v
+      | v <- findVars ctx Proxy
+      , case lookupVar ctx v of
+          MEither (Left _)            -> False
+          MEither (Right (MCursor c)) ->
+            Map.member (Model.cursorID c) (Model.cursors st)
+      ]
+
+    genBlobRefsVar = QC.elements blobRefsVars
+
+    blobRefsVars :: [Var h (V.Vector (WrapBlobRef h IO blob))]
+    blobRefsVars = fmap (mapGVar (OpComp OpLookupResults)) lookupResultVars
+                ++ fmap (mapGVar (OpComp OpQueryResults))  queryResultVars
       where
-        vars1 = findVars ctx (Proxy @(Either Model.Err (V.Vector (R.LookupResult v (WrapBlobRef h IO blob)))))
-        vars2 = findVars ctx (Proxy @(Either Model.Err (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob)))))
-
-        fromLookupResults ::
-             Var h (Either Model.Err (V.Vector (R.LookupResult v (WrapBlobRef h IO blob))))
-          -> Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))
-        fromLookupResults = mapGVar (\op -> OpRight `OpComp` OpLookupResults `OpComp` OpFromRight `OpComp` op)
-
-        fromQueryResults ::
-             Var h (Either Model.Err (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob))))
-          -> Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))
-        fromQueryResults = mapGVar (\op -> OpRight `OpComp` OpQueryResults `OpComp` OpFromRight `OpComp` op)
-
-    withoutVars :: [(Int, Gen (Any (LockstepAction (ModelState h))))]
-    withoutVars = [
-          (5, Some . New (PrettyProxy :: PrettyProxy (k, v, blob)) <$> QC.arbitrary)
-        , (3, fmap Some $ Open @k @v @blob <$> genSnapshotName)
-        , (1, fmap Some $ DeleteSnapshot <$> genSnapshotName)
-        , (1, pure $ Some ListSnapshots)
+        lookupResultVars :: [Var h (V.Vector (R.LookupResult  v (WrapBlobRef h IO blob)))]
+        queryResultVars  :: [Var h (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob)))]
+
+        lookupResultVars = fromRight <$> findVars ctx Proxy
+        queryResultVars  = fromRight <$> findVars ctx Proxy
```

Comment:
What would be nice to have at some point is that we could mix and match blob references that come from different lookups/cursor reads. There is already a TODO in the module header, but I just thought I'd mention it again :stuck_out_tongue:

Reply:
Yes I agree... at some point :smiley: